Fix PNPM issues with private registries

deivid-rodriguez opened the pull request:
This PR fixes two issues:

* Dependabot can't infer the proper registry for each dependency by looking at registry urls in pnpm-lock.yaml.

This is handy because developers don't need to commit a `.npmrc` file and matches what we already support for NPM and Yarn.

* Dependabot incorrectly tries to infer the proper registry for each dependency by looking at presence in the registry, even when registries & scopes are explicitly configured in `.npmrc`.

This is a good last resort fallback, but it should not be used if registries & scopes have been explicitly configured.

Fixes https://github.com/dependabot/dependabot-core/issues/8242.
Fixes performance issues introduced by https://github.com/dependabot/dependabot-core/pull/8094.
Likely to fix https://github.com/dependabot/dependabot-core/issues/8008.

yeikel commented:
Thank you for tagging me! 

I am not too familiar with the ecosystem but I understand the problem. Any chance you can add a spec?

deivid-rodriguez commented:
Actually I meant to ping @mburumaxwell 🤦‍♂️, sorry about that. Thanks for chiming in though! Yes, this is a draft because I plan to add some specs 👍.

deivid-rodriguez commented:
@dancallaghan This is now ready, I'll deploy it next week!

deivid-rodriguez commented on npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb:
```diff
@@ -49,16 +49,22 @@ def self.central_registry?(registry)
-
-          scope = dependency_name.split("/").first
-          scoped_registry(scope)
+          explicit_registry_from_rc(dependency_name) || global_registry
```
I did this for backwards compatibility with other usages of `RegistryFinder#registry_from_rc`. I think it's used just once and could probably use the changed behavior too, I'll look into what would mean to change it.

It would make the code a bit simpler and may fix additional issues, so worth looking into it.

deivid-rodriguez replied:
I had a look at this but decided it was not worth exploring further right now.

yeikel commented on npm_and_yarn/helpers/lib/pnpm/lockfile-parser.js:
```diff
@@ -33,6 +33,7 @@ function nameVerDevFromPkgSnapshot(depPath, pkgSnapshot, projectSnapshots) {
 
   let specifiers = [];
   let aliased = false;
+  let resolved = pkgSnapshot.resolution.tarball;
```
nit: Is there any reason reason to use `let` as opposed to `const`?

yeikel commented on npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb:
```diff
@@ -208,34 +215,40 @@ def unique_registries(registries)
             next if Regexp.last_match[:registry].include?("${")
 
-            return @global_registry = Regexp.last_match[:registry].strip
+            return @configured_global_registry = Regexp.last_match[:registry].strip
```
Would it make sense to extract `Regexp.last_match[:registry].strip` to a method?

yeikel commented on npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb:
```diff
@@ -279,6 +292,12 @@ def parsed_yarnrc_yml
           @parsed_yarnrc_yml = YAML.safe_load(yarnrc_yml_file.content)
         end
+
+        def normalize_configured_registry(url)
+          url.sub(%r{/+$}, "")
+             .sub(%r{^.*?//}, "")
+             .gsub(/\s+/, "%20")
+        end
```
If you have a spec testing this then please point me to it

deivid-rodriguez replied:
I added this to respect the format currently returned by `RegistryFinder#registry`. The format is a URL with no protocol, no trailing slashes, and escaped espaces. However, they may appear in other format in rc files, so needs to be normalized. There are many specs that fail if you don't do this.

Note that I did not add this logic, it was already there, I just extracted it to method so that I can reuse it.

deivid-rodriguez commented:
Thanks for reviewing @yeikel, hopefully I addressed your comments.

@dancallaghan If you're able to try this again, I appreciate. I introduced a few refactoring so it'd be good to know I did not break anything!

dancallaghan commented:
Thanks @deivid-rodriguez

I've tested locally with the latest changes on our use case and it's working as expected 👍

BPScott commented:
@deivid-rodriguez, It seems like the changes to the lockfile in pnpm v9 mean these improvements no longer apply.

The performance issues caused by dependabot checking non-standard repositories for all packages rather than the ones within configured namespaces have resurfaced

deivid-rodriguez commented:
I no longer actively contribute to this repo, so I won't be able to help right now, but thanks for letting me know!
